substrate/frame/beefy/src/equivocation.rs

```diff
@@ -166,19 +188,42 @@ where
 	fn check_evidence(
 		evidence: EquivocationEvidenceFor<T>,
 	) -> Result<(), TransactionValidityError> {
-		let (equivocation_proof, key_owner_proof) = evidence;
-
-		// Check the membership proof to extract the offender's id
-		let key = (BEEFY_KEY_TYPE, equivocation_proof.offender_id().clone());
-		let offender = P::check_proof(key, key_owner_proof).ok_or(InvalidTransaction::BadProof)?;
-
-		// Check if the offence has already been reported, and if so then we can discard the report.
-		let time_slot = TimeSlot {
-			set_id: equivocation_proof.set_id(),
-			round: *equivocation_proof.round_number(),
+		let (offenders, key_owner_proofs, time_slot) = match &evidence {
+			EquivocationEvidenceFor::VoteEquivocationProof(equivocation_proof, key_owner_proof) => {
+				// Check if the offence has already been reported, and if so then we can discard the
+				// report.
+				let time_slot = TimeSlot {
+					set_id: equivocation_proof.set_id(),
+					round: *equivocation_proof.round_number(),
+				};
+				(vec![equivocation_proof.offender_id()], vec![key_owner_proof.clone()], time_slot)
+			},
+			EquivocationEvidenceFor::ForkEquivocationProof(
+				equivocation_proof,
+				key_owner_proofs,
+			) => {
+				// Check if the offence has already been reported, and if so then we can discard the
+				// report.
+				let time_slot = TimeSlot {
+					set_id: equivocation_proof.set_id(),
+					round: *equivocation_proof.round_number(),
+				};
+				let offenders = equivocation_proof.offender_ids(); // clone data here
+				(offenders, key_owner_proofs.to_owned(), time_slot)
+			},
 		};
 
-		if R::is_known_offence(&[offender], &time_slot) {
+		// Validate the key ownership proof extracting the id of the offender.
+		let offenders = offenders
```

Comment:
Should we also check for duplicate authorities + signatures here ? If not someone can inflate the size of the proof artificially, while being refunded => DoS vector ?

Reply:
https://github.com/lederstrumpf/polkadot-sdk/blob/e7718fc8e75ed7133b3c06061374f32beefa7b8f/substrate/frame/grandpa/src/equivocation.rs#L167-L171
https://github.com/lederstrumpf/polkadot-sdk/blob/e7718fc8e75ed7133b3c06061374f32beefa7b8f/substrate/frame/beefy/src/equivocation.rs#L237-L241

Equivocation reporting takes a lax stance: equivocations are reported if *any* of the included offences are new.

> Should we also check for duplicate authorities + signatures here ?
> If not someone can inflate the size of the proof artificially, while being refunded => DoS vector

Firstly, I'm not convinced the problem should be handled here, but rather be the responsibility of the `offences` pallet since the problem is generic to equivocation reports, so shouldn't be solved at multiple call sites in parallel.

Secondly, when the report is triaged, the offenders are iterated over and only if an offender has not been added to the `ReportIndexStorage` yet, the report is added to `Reports`, and the offender added to `ReportIndexStorage`:
https://github.com/lederstrumpf/polkadot-sdk/blob/master/substrate/frame/offences/src/lib.rs#L175-L187

So even if the offender is duplicated in the same report (with distinct signatures), the reward would only be accounted for once. This does leave the door open for duplicated submissions nonetheless, but there's no reward incentive for doing so.

Reply:
I don't mean for us to handle duplicate offence reports. That's indeed the responsibility of the offences pallet.

I mean that if someone sends a fork equivocation report with duplicate authorities, signatures, and key ownership proofs, for example: (A, B, C, C, C, C, C ... 1 million times C) and the report is ok, the system will spend compute cycles in checking C 1 million times while the sender will be refunded for the cost of the extrinsic. Which seems like some kind of DoS. Or do we have a limit on the number of authorities in a report somewhere ?

Reply:
please move all review conversations to #1903 ; this PR is closed